### PR TITLE
Include external config

### LINKED ISSUE #318: feature: including external docker-compose.yml
### The Problem

I'd like to use fig for acceptance testing of services

In the case of a single service with a single level of dependencies this is easy (ex: a webapp with a database). As soon as the dependency tree grows to depth > 1, it gets more complicated (ex: service-a requires service-b which requires a database). Currently I'd have to specify the service dependency tree in each `fig.yml`.  This is not ideal because as the tree grows, we end up with a lot of duplication, and having to update a few different `fig.yml`'s in different projects when a service adds/changes a dependency is not great.
### Proposed Solution

Support an `include` section in the `fig.yml` which contains url/paths to other `fig.yml` files. These files would be included in a `Project` so that the top-level `fig.yaml` can refer to them using `<proejct>_<service>`.
#### Example config

``` yaml
include:
    # link using http
    servicea:
        http: http://localhost:8080/projects/servicea/fig.yml
    # local paths
    serviceb:
        path: ../projects/serviceb/fig.yml

webapp:
    ...
    links:
        - servicea_webapp
        - serviceb_webapp
```

### PR DESCRIPTION
**Pull request by dnephin**
Resolves #318

Adds support for including other `fig.yml` files by url (or relative path).  

Requires an additional `fig.yml` key of `project-config` to support additional configuration.  I would see this being used for things like project name as well.

Feedback welcome

### CODE REVIEW

**Comment by aanand** [comment]
OK, just had a read over. Promising stuff. Thoughts, in decreasing order of (aha) import:

1. This lets you import an entire set of services from another project. Could it be easily extended so that you could import just a defined subset of services? Would people want to do that?

2. Suppose the file format changes in future such that it contains more than just service definitions. Would users want to be able to import other things from external projects, not just services? Could we make the import mechanism sufficiently generic to support that? (Sorry, I'll try and come up with a good example for this)

3. Resolution of URLs is really cool, but introduces a bunch of new concerns (most obviously, caching and the management thereof). Could an initial version leave it out and still be useful?

4. Bikeshedding the `project-config` key for a bit: could we make it stand out more? I was wondering once about a key for global not-service config being all-uppercase, e.g. `PROJECT`. (Also, we're using underscores everywhere else, so probably shouldn't introduce dashes here).

4. I see you're frequently using `pop()` instead of `get()` on dictionaries. What's the rationale for mutating loaded data like that? It feels a bit weird to me.

5. Spotted a couple of typos (`pojrectb_webapp`, `proejctc`).

**Comment by dnephin** [comment]
Thanks for the feedback!

1. I could see the import section also taking a specific list of service names, and the external project only including the named services (and their dependencies). It makes things a bit more complicated, but it might be nice for the case where external configs contain a bunch of extra things.  Right now I'm using a small tool which rewrites a fig.yml to replace `build` directives with `image`. Since `build` isn't support from external projects. It might make sense to support this type of filtering in that re-write-fig.yml tool instead (those operations seem related to me). If that's an approach we want to explore, I will work on getting the source up on github.

2. I suspect importing other settings from the config would not be useful, but it's hard to say until we actually have other things in the config. 

3. Yes, totally agree, the fetching of remote config by url does add a bunch of new concerns. I think for many use-cases it would still be quite useful without this.  The fetch of external files could always be done with a curl/s3cmd before running `fig`. We could exclude it from an initial release.

4. `project-config` is easy enough to change. If the config format changes to include a section for services (like the compose proposal suggested), it can be removed entirely, and the top-level key would just be `include`.  It could also change it to `PROJECT` if that works better. The reason I use a dash is to make sure it's backwards compatible. Someone could potentially have a service with the name `PROJECT` which would break when they upgrade (however I guess that is pretty unlikely).

5. Just backwards compatibility (if other code does validation on the expected keys).  I think it should be fine to change to a `get()`

File: fig/includes.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    url = urlparse(url)
+    return url if url.scheme else url._replace(scheme='file')
+
+
+def read_config(content):
+    return yaml.safe_load(content)
+
+
+def get_project_from_file(url):
+    # Handle urls in the form file://./some/relative/path
```
Comment (sdake) [inline fig/includes.py]: Does this handle files in local directories?  If not, it would be great if it did :)

File: fig/includes.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    except requests.exceptions.RequestException as e:
+        raise FetchExternalConfigError("Failed to include %s: %s" % (
+            url.geturl(), e))
+    return read_config(response.text)
+
+
+# Return the connection from a function, so it can be mocked in tests
+def get_boto_conn():
+    # Local import so that boto is only a dependency if it's used
+    import boto.s3.connection
```
Comment (sdake) [inline fig/includes.py]: nice solution boto is poorly maintained.

**Comment by sdake** [comment]
@dnephin reviewed this patch looks solid - although needs a rebase.  Can we have this in master soon? :)

File: fig/includes.py (hunk 1 above)
Comment (dnephin) [inline fig/includes.py]: Yup, I believe that relative paths were working.

I should say that this implementation will probably change quite a bit. It was more of a proof-of-concept

**Comment by zaneb** [comment]
Code looks good to me.

**Comment by omerzimp** [comment]
+1 on needing to rebase this.

File: docs/yml.md [hunk 3]
```diff
@@ -199,3 +199,42 @@ privileged: true
...
+using the specified cache settings (defaults to a path of ~/.fig-cache/ with
+a ttl of 5 minutes).
+
+Example:
+
+```yaml
+
+project-config:
+
+    include:
```
Comment (erichonkanen) [inline docs/yml.md]: could this also use a git repository with a docker-compose.yml file in it?

File: docs/yml.md [hunk 4]
```diff
@@ -199,3 +199,42 @@ privileged: true
...
+            url: 's3://bucket/path/to/key/projecta.yml'
+        projectb:
+            url: 'http://example.com/projectb/fig.yml'
+        projectc:
+            url: './path/to/projectc/fig.yml'
+
+    # This section is optional, below are the default values
+    cache:
+        enable: True
+        path: ~/.fig-cache/
```
Comment (erichonkanen) [inline docs/yml.md]: should this be ~/.compose/ ? isn't fig deprecated

**Comment by scottbelden** [comment]
Any progress on trying to rebase this and get it in? Would love to be able to use this.

**Comment by dnephin** [comment]
@scottbelden Not yet, it still needs a lot of work, and probably the removal of some features for an initial release.  

I think I might try and extract this code as a separate tool that can be used to pipe the stdin once #1488 is merged. Once that's ready, I'll leave a comment in this PR.

**Comment by dnephin** [comment]
I've implemented this as an external pre-processing tool at https://github.com/dnephin/compose-addons#dcao-include

If anyone is interested, please check out the docs. Any feedback can be contributed by opening an issue on that repo (and would be appreciated).

**Pull request closed without merging**